src/IR/DesugarPattern.hs

```diff
@@ -241,14 +250,21 @@ desugarMatchCons (u : us) qs def = do
             argsTyps = argsType cinfo
             makeVar (i, t) = do
                 return (I.Var i t)
-        us'  <- mapM makeVar $ zip newVars argsTyps
-        body <- desugarMatch
-            (us' ++ us)
-            [ (as' ++ as, e) | ((I.AltData _ as') : as, e) <- qs' ]
-            def
-        let makeBinder (I.Var vid _) = I.AltDefault $ Just vid
+            makeBinder (I.Var vid _) = I.AltDefault $ Just vid
             makeBinder _             = error "can't happen"
-        return (I.AltData dcon (map makeBinder us'), body)
+        us' <- mapM makeVar $ zip newVars argsTyps
+        case qs' of
+            [] ->
+                return
+                    ( I.AltData dcon (map makeBinder us')
+                    , I.NoExpr (I.extract def)
+                    )
+            _ -> do
+                body <- desugarMatch
+                    (us' ++ us)
+                    [ (as' ++ as, e) | ((I.AltData _ as') : as, e) <- qs' ]
+                    def
+                return (I.AltData dcon (map makeBinder us'), body)
```

Comment:
I think this change caused some problems, because it sets the DCon that are not in the pats directly to NoExpr, which is not desired